### ISSUE TITLE
Android c++ build under Linux links against pthread which isn't available

### ISSUE THREAD

**Issue by MarkaRagnos0815** [bazelbuild/bazel#7983]
### Description of the problem / feature request:

I am using the following command to build a android so under Linux
`` 
bazel build <target> --compilation_mode=opt --crosstool_top=@androidndk//:default_crosstool --host_crosstool_top=@bazel_tools//tools/cpp:toolchain --cpu=armeabi-v7a

error:
external/androidndk/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lpthread
external/androidndk/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lpthread
`` 

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

- setup a cc_binary to build a shared object (.so extension)
- try to build using android crosstool according to commanline above (use ndk r16b)

### What operating system are you running Bazel on?

- ubuntu 18.04 (in may case ubuntu shell in windows 10)

### What's the output of `bazel info release`?

- release 0.24.1

###  Have you found anything relevant by searching the web?

- no

### Any other information, logs, or outputs that you want to share?

- command line looks like when build with "-s"
`` 
external/androidndk/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang -shared -o  ... 
external/androidndk/ndk/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/libgnustl_static.a external/androidndk/ndk/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/libsupc++.a -llog -lc -lpthread -lm -lpthread -lm -gcc-toolchain external/androidndk/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64 -no-canonical-prefixes -target armv7-none-linux-androideabi -Wl,--fix-cortex-a8 -static-libgcc '--sysroot=external/androidndk/ndk/platforms/android-26/arch-arm
``

**Comment by jin** [comment]
Can you please provide a repro project? `-lpthread` is not a default linkopt in the NDK crosstool.

**Comment by MarkaRagnos0815** [comment]
I am currently not able to reproduce it with a simple project.
The project is really complex and contains about 25 static libraries and since it does not use linker input files I need to test it using LSW (command line to long in windows and LSW performance is really bad, 50sec per file to compile)

I will try to provide a sample but it will take some time.
But maybe the reason could be a cc_proto_library (just a guess)

**Comment by MarkaRagnos0815** [comment]
[test.zip](https://github.com/bazelbuild/bazel/files/3130099/test.zip)
see attached zip

**Comment by mchinen** [comment]
The issue seems to be that bazel adds linkopts from the targets built with the host (such as protoc, for the protobuf dependency) into the linkopts for the target.  It would seem that android_binary() somehow resolves this in a way that cc_binary does not.  I captured this in a question here.
https://stackoverflow.com/questions/66394377/bazel-host-x86-linux-linkopts-propogating-to-target-android

**Comment by cpsauer** [comment]
I'd worked around this a while back, and I just saw this issue. One hits this problem whenever the NDK tries to use an external library that (sensibly) tries to link against lpthread because it uses pthread headers.

[You run into this trying to use rules_boost on Android, for example](https://github.com/nelhage/rules_boost/issues/165), and it's trickier to work around because bazel platform support is in an incomplete state.  

**Re the "more data needed" tag:** 
Here's what's going on: As per the stackoverflow post and that boost issue above, the linker is getting the -lpthread flag not from the user's code but from an external dependency. The Android NDK is weird in that it bundles pthread's symbols into libc, rather than having a separate libpthread like other posix-type platforms. This breaks compatibility build scripts designed to work across platforms; -lpthread can't find a matching library because android has put its symbols under a different name. 

**Suggested fix:**
IMO the most elegant fix for Bazel is to make Android's NDK behave more like one would expect of a Linux system. A more uniform platform makes code reuse easier and reduces suprises like these! 

The easiest way to do that is to make -libpthread a no op on Android. This doesn't break anything, because the symbols you'd expect to receive are already included in by libc, which the NDK already implicitly linked against.  As mentioned in the boost issue, I did this by having a dummy libpthread with no symbols, but there are probably more elegant fixes if we can change Bazel source.  

[You're also likely to hit some other NDK linker flag inconsistencies](https://github.com/bazelbuild/bazel/issues/10837) that compound the issue of debugging this; might be good to fix those at the same time.

Happy to contribute my workaround as an external dep if some others are running into this, too. LMK.

-CS
